### PR TITLE
ARROW-14594: [R] Enable snappy+lz4 by default

### PR DESCRIPTION
**Pull request by jonkeane**
Since they are used by default when they are available

### CODE REVIEW

**Comment by jonkeane** [comment]
@github-actions crossbow submit -g r

**Comment by jonkeane** [comment]
Turns out this exposes a sanitizer [issue in snappy](https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=15095&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=d9b15392-e4ce-5e4c-0c8c-b69645229181&l=4006):

```
/tmp/RtmpR2qogf/file50379aa3da/snappy_ep-prefix/src/snappy_ep/snappy.cc:95:19: runtime error: load of misaligned address 0x60a0000052c1 for type 'const uint32', which requires 4 byte alignment
0x60a0000052c1: note: pointer points here
 00 00 13  00 00 00 00 00 00 35 40  66 66 66 66 66 66 35 40  9a 99 99 99 99 19 32 40  33 33 33 33 33
              ^ 
    #0 0x7f376be87331 in snappy::internal::CompressFragment(char const*, unsigned long, char*, unsigned short*, int) (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x30bc2331)
    #1 0x7f376be8bc82 in snappy::Compress(snappy::Source*, snappy::Sink*) (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x30bc6c82)
    #2 0x7f376be91439 in snappy::RawCompress(char const*, unsigned long, char*, unsigned long*) (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x30bcc439)
    #3 0x7f3760bc6e29 in arrow::util::internal::(anonymous namespace)::SnappyCodec::Compress(long, unsigned char const*, long, unsigned char*) (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x25901e29)
    #4 0x7f375c2b7176 in parquet::SerializedPageWriter::Compress(arrow::Buffer const&, arrow::ResizableBuffer*) (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x20ff2176)
    #5 0x7f375c356565 in parquet::SerializedPageWriter::WriteDictionaryPage(parquet::DictionaryPage const&) (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x21091565)
    #6 0x7f375c2a8560 in parquet::TypedColumnWriterImpl<parquet::PhysicalType<(parquet::Type::type)5> >::WriteDictionaryPage() (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x20fe3560)
    #7 0x7f375c25e021 in parquet::ColumnWriterImpl::Close() (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x20f99021)
    #8 0x7f375bce64ad in parquet::arrow::FileWriterImpl::WriteColumnChunk(std::shared_ptr<arrow::ChunkedArray> const&, long, long) (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x20a214ad)
    #9 0x7f375bcb6f52 in parquet::arrow::FileWriterImpl::WriteTable(arrow::Table const&, long) (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x209f1f52)
    #10 0x7f375b5f19f2 in arrow::dataset::ParquetFileWriter::Write(std::shared_ptr<arrow::RecordBatch> const&) (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x2032c9f2)
    #11 0x7f375b743040 in arrow::internal::FnOnce<void ()>::FnImpl<std::_Bind<arrow::detail::ContinueFuture (arrow::Future<unsigned long>, arrow::dataset::internal::(anonymous namespace)::DatasetWriterFileQueue::WriteNext()::{lambda()#1})> >::invoke() (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x2047e040)
    #12 0x7f37608b3ad0 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<arrow::internal::ThreadPool::LaunchWorkersUnlocked(int)::{lambda()#1}> > >::_M_run() (/usr/local/RDsan/lib/R/site-library/arrow/libs/arrow.so+0x255eead0)
    #13 0x7f379537ade3  (/usr/lib/x86_64-linux-gnu/libstdc++.so.6+0xd6de3)
    #14 0x7f3795e12608 in start_thread /build/glibc-eX1tMB/glibc-2.31/nptl/pthread_create.c:477
    #15 0x7f3795d37292 in __clone (/usr/lib/x86_64-linux-gnu/libc.so.6+0x122292)
```

https://dev.azure.com/ursacomputing/crossbow/_build/latest?definitionId=1&branchName=actions-1093-azure-test-ubuntu-18.04-r-sanitizer

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-ubuntu-18.04-r-sanitizer

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-ubuntu-18.04-r-sanitizer

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by jonkeane** [comment]
The centos 7 error is real, though I don't understand why it's happening. It seems to be hitting [one of these two blocks](https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/compression_snappy.cc#L47-L59)

I've tried saving a snappy encoded parquet from inside of a centos machine with this installed and reading it on my local (macOS) and that works _just fine_. I also did the reverse (saved a parquet file on my local and tried to read it in in the centos7 container, and that errors with the same corruption message. 

So it doesn't look like the file is truly corrupt, but something is odd with our code on centos 7  specifically(?) 

Error:
```
  > tf <- tempfile()
  > on.exit(unlink(tf))
  > write_parquet(mtcars, tf)
  > df <- read_parquet(tf, col_select = starts_with("d"))
  Error: IOError: Corrupt snappy compressed data.
```

**Comment by jonkeane** [comment]
@github-actions autotune

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest test-r-rhub-ubuntu-gcc-release-latest test-r-rocker-r-base-latest test-r-rstudio-r-base-3.6-bionic test-r-rstudio-r-base-3.6-centos7-devtoolset-8 test-r-rstudio-r-base-3.6-centos8 test-r-rstudio-r-base-3.6-opensuse15 test-r-rstudio-r-base-3.6-opensuse42

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest test-r-rhub-ubuntu-gcc-release-latest test-r-rocker-r-base-latest test-r-rstudio-r-base-3.6-bionic test-r-rstudio-r-base-3.6-centos7-devtoolset-8 test-r-rstudio-r-base-3.6-centos8 test-r-rstudio-r-base-3.6-opensuse15 test-r-rstudio-r-base-3.6-opensuse42 test-ubuntu-18.04-r-sanitizer

**Comment by jonkeane** [comment]
@github-actions crossbow submit -g cpp

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -579,10 +579,18 @@ endif()
 if(DEFINED ENV{ARROW_SNAPPY_URL})
   set(SNAPPY_SOURCE_URL "$ENV{ARROW_SNAPPY_URL}")
 else()
-  set_urls(SNAPPY_SOURCE_URL
-           "https://github.com/google/snappy/archive/${ARROW_SNAPPY_BUILD_VERSION}.tar.gz"
-           "https://github.com/ursa-labs/thirdparty/releases/download/latest/snappy-${ARROW_SNAPPY_BUILD_VERSION}.tar.gz"
-  )
+  if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS
```
Comment (nealrichardson) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:582]: Can you leave a comment explaining this? (I see you have one in versions.txt but would be nice to explain something here too.)

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 2]
```diff
@@ -579,10 +579,19 @@ endif()
...
+                                              "4.9")
+    # There is a bug in GCC < 4.9 with Snappy 1.1.9, so revert to 1.1.8 "SNAPPY_OLD" for those
```
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:584]: "a bug" is a bit frustrating to the reader, can you point to the discussion or JIRA?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 3]
```diff
@@ -579,10 +579,19 @@ endif()
...
+    set_urls(SNAPPY_SOURCE_URL
+             "https://github.com/google/snappy/archive/${ARROW_SNAPPY_OLD_BUILD_VERSION}.tar.gz"
+             "https://github.com/ursa-labs/thirdparty/releases/download/latest/snappy-${ARROW_SNAPPY_OLD_BUILD_VERSION}.tar.gz"
+    )
+  else()
+    set_urls(SNAPPY_SOURCE_URL
+             "https://github.com/google/snappy/archive/${ARROW_SNAPPY_BUILD_VERSION}.tar.gz"
+             "https://github.com/ursa-labs/thirdparty/releases/download/latest/snappy-${ARROW_SNAPPY_BUILD_VERSION}.tar.gz"
+    )
+  endif()
```
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:594]: Can you also set something like `SNAPPY_BUILD_CHECKSUM` here...

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 4]
```diff
@@ -971,9 +980,18 @@ macro(build_snappy)
...
+      ${EP_COMMON_CMAKE_ARGS}
+      -DCMAKE_INSTALL_LIBDIR=lib
+      -DSNAPPY_BUILD_TESTS=OFF
+      -DSNAPPY_BUILD_BENCHMARKS=OFF
       "-DCMAKE_INSTALL_PREFIX=${SNAPPY_PREFIX}")
 
+  # There is a bug in GCC < 4.9 with Snappy 1.1.9, so revert to 1.1.8 "SNAPPY_OLD" for those
+  if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS
+                                              "4.9")
+    set(ARROW_SNAPPY_BUILD_SHA256_CHECKSUM ${ARROW_SNAPPY_OLD_BUILD_SHA256_CHECKSUM})
```
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:992]: ... rather than here?

**Comment by jonkeane** [comment]
@github-actions crossbow submit -g r

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-linux-as-cran

**Pull request closed without merging**